### ISSUE TITLE
Data Parallel

### ISSUE THREAD

**Issue by yhgon** [facebookresearch/SING#3]
Amazing works !

I found some issues  with `--parallel` 
my environment is pytorch 1.0.0a0 with CUDA 10  docker images from `hryu/pytorch:t2-ngc-18.11`  

- 1. Memory Alloc. 
I could not increase batch_size because first GPU alloc more  memory. even other GPUs have more room. 
```
| N/A   46C    P0   103W / 350W |  29200MiB / 32510MiB 
| N/A   44C    P0    82W / 350W |  11244MiB / 32510MiB  
| N/A   52C    P0    87W / 350W |  11244MiB / 32510MiB  
```
- 2. scale in AE train.  
data parallel works in my machine but scale factor is not good as I expect. ( AE 85 ex/s in single GPU)    
```
AE : 231212/231212 [06:43<00:00, 570.59ex/s
SE : 131072/231212 [01:08<00:50, 1983.36ex/s
```

- 3. GPU is idle for every iter in train process.
I found the reason of the scale issue occur from this issue. 
during GPU idle time, CPU utilization increase 3~4000% with 40 Core CPU. 
When I increase the batchsize, the idle interval increase.  
```
| N/A   48C    P0   273W / 350W |  19492MiB / 32510MiB |    100%      Default |
| N/A   47C    P0   252W / 350W |  16434MiB / 32510MiB |    100%      Default |
| N/A   56C    P0   301W / 350W |  16434MiB / 32510MiB |    100%      Default |
| N/A   48C    P0   313W / 350W |  16434MiB / 32510MiB |    100%      Default |
| N/A   42C    P0    83W / 350W |  19492MiB / 32510MiB |      0%      Default |
| N/A   41C    P0    81W / 350W |  16434MiB / 32510MiB |      0%      Default |
| N/A   49C    P0    87W / 350W |  16434MiB / 32510MiB |      0%      Default |
| N/A   48C    P0    85W / 350W |  16434MiB / 32510MiB |      0%      Default |
| N/A   41C    P0    83W / 350W |  19492MiB / 32510MiB |      0%      Default |
| N/A   40C    P0    81W / 350W |  16434MiB / 32510MiB |      0%      Default |
| N/A   48C    P0    86W / 350W |  16434MiB / 32510MiB |      0%      Default |
| N/A   47C    P0    84W / 350W |  16434MiB / 32510MiB |      0%      Default |
| N/A   40C    P0   179W / 350W |  19492MiB / 32510MiB |      0%      Default |
```

**Comment by adefossez** [comment]
@yhgon I am planning on using DDP for my future work but at the time DP was the most obvious choice at the time.
I have not tried to make the training scale to such a large level of GPUs. In my experiments 4 GPUs was giving the ideal speedup. Using DDP might help to scale further.

I now realize that there are some design mistakes in the way I used DataParallel. In particular, the generation of the waveform is parallelized and then gathered on GPU 0 and the loss is computed only on GPU 0. This explain the increased memory usage for GPU0 as well as the some suboptimal usage of the GPUs. Changing that would require a bit of work but I don't currently have a lot of time to do so.
